velox/substrait/SubstraitToVeloxExpr.cpp

```diff
@@ -16,6 +16,128 @@
+  return literal.i32();
+}
+
+template <>
+int64_t getLiteralValue(const ::substrait::Expression::Literal& literal) {
+  return literal.i64();
+}
+
+template <>
+double getLiteralValue(const ::substrait::Expression::Literal& literal) {
+  return literal.fp64();
+}
+
+template <>
+float getLiteralValue(const ::substrait::Expression::Literal& literal) {
+  return literal.fp32();
+}
+
+template <>
+bool getLiteralValue(const ::substrait::Expression::Literal& literal) {
+  return literal.boolean();
+}
+
+template <>
+uint32_t getLiteralValue(const ::substrait::Expression::Literal& literal) {
+  return literal.i32();
+}
+
+ArrayVectorPtr makeArrayVector(const VectorPtr& elements) {
+  BufferPtr offsets = allocateOffsets(1, elements->pool());
+  BufferPtr sizes = allocateOffsets(1, elements->pool());
+  sizes->asMutable<vector_size_t>()[0] = elements->size();
+
+  return std::make_shared<ArrayVector>(
+      elements->pool(),
+      ARRAY(elements->type()),
+      nullptr,
+      1,
+      offsets,
+      sizes,
+      elements);
+}
+
+ArrayVectorPtr makeEmptyArrayVector(memory::MemoryPool* pool) {
+  BufferPtr offsets = allocateOffsets(1, pool);
+  BufferPtr sizes = allocateOffsets(1, pool);
+  return std::make_shared<ArrayVector>(
+      pool, ARRAY(UNKNOWN()), nullptr, 1, offsets, sizes, nullptr);
+}
+
+template <TypeKind kind>
+void listLiteralToFlatVector(
+    const ::substrait::Expression::Literal& listLiteral,
+    const VectorPtr& vector,
+    vector_size_t& index) {
+  using T = typename TypeTraits<kind>::NativeType;
+  auto flatVector = vector->as<FlatVector<T>>();
+  auto typeCase = listLiteral.literal_type_case();
+  if (typeCase == ::substrait::Expression_Literal::LiteralTypeCase::kList) {
+    for (auto child : listLiteral.list().values()) {
+      listLiteralToFlatVector<kind>(child, vector, index);
+    }
+    return;
+  }
+
+  if constexpr (std::is_same_v<T, StringView>) {
+    if (typeCase == ::substrait::Expression_Literal::LiteralTypeCase::kString) {
+      flatVector->set(index++, StringView(listLiteral.string()));
+    }
+    if (typeCase ==
+        ::substrait::Expression_Literal::LiteralTypeCase::kVarChar) {
+      flatVector->set(index++, StringView(listLiteral.var_char().value()));
+    }
+  } else if constexpr (
+      std::is_same_v<T, Timestamp> || std::is_same_v<T, Date> ||
+      std::is_same_v<T, IntervalDayTime>) {
+  } else {
+    if (typeCase == ::substrait::Expression_Literal::LiteralTypeCase::kNull) {
+      vector->setNull(index++, true);
+    } else {
+      flatVector->set(index++, getLiteralValue<T>(listLiteral));
+    }
+  }
+}
+
+template <TypeKind kind>
+VectorPtr constructFlatVector(
+    const ::substrait::Expression::Literal& listLiteral,
+    const vector_size_t size,
+    const TypePtr& type,
+    memory::MemoryPool* pool) {
+  VELOX_CHECK(type->isPrimitiveType());
+  auto vector = BaseVector::create(type, size, pool);
+
+  vector_size_t index = 0;
```

Comment:
The use of 'index' is confusing? Why can't we have a simple loop?

Reply:
It's really confusing here. But I don't think the loop is very suitable here. because listLiteralToFlatVector will call listLiteralToFlatVector itself here. 
I can use "0“ to replace index directly. like this:
`listLiteralToFlatVector<kind>(listLiteral, vector, 0);`

Reply:
@zhejiangxiaomai Yes, this code is unnecessarily confusing. It might be clearer to use a loop like so:

```
  vector_size_t index = 0;
  for (auto child : listLiteral.list().values()) {
    setLiteralValue(child, flatVector, index++);
  }

...
template <typename T>
void setLiteralValue(
    const ::substrait::Expression::Literal& literal,
    FlatVector<T>* vector,
    vector_size_t index) {
  if (literal.has_null()) {
    vector->setNull(index, true);
  } else if constexpr (std::is_same_v<T, StringView>) {
    const auto typeCase = literal.literal_type_case();
    if (literal.has_string()) {
      vector->set(index, StringView(literal.string()));
    } else if (literal.has_var_char()) {
      vector->set(index, StringView(literal.var_char().value()));
    } else {
      VELOX_FAIL("Unexpected string literal");
    }
  } else {
    vector->set(index, getLiteralValue<T>(literal));
  }
}
```

See complete code in https://github.com/facebookincubator/velox/compare/main...mbasmanova:velox-1:support_list

Reply:
Thanks for your opinion. I will follow this way.